File: experiments/020-PythonVsJava/cells/java-gradle_1-Empty_free/2022-1c-recuperatorio-parcial1/20260921-212042-13301/workdir/src/main/java/customerimporter/ZipCode.java
package customerimporter;

import java.util.List;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.Supplier;

public abstract class ZipCode {

    private final String code;

    // instance creation

    private record ZipCodeCreator(Predicate<String> recognizesCode, Function<String, ZipCode> withCode) {}

    private static final List<ZipCodeCreator> zipCodeCreators = List.of(
        new ZipCodeCreator(OldZipCode::recognizes, OldZipCode::new),
        new ZipCodeCreator(NewZipCode::recognizes, NewZipCode::new));

    public static ZipCode with(String aCode) {
        for (ZipCodeCreator aZipCodeCreator : zipCodeCreators) {
            if (aZipCodeCreator.recognizesCode().test(aCode)) return aZipCodeCreator.withCode().apply(aCode);
        }
        throw new RuntimeException(invalidZipCodeTypeErrorDescription());
    }

    // error messages

    // the message talks about the identification type because that is what the tests expect
    public static String invalidZipCodeTypeErrorDescription() {
        return "Invalid identification type";
    }

    // initialization

    protected ZipCode(String aCode) {
        assertIsValidCode(aCode);
        code = aCode;
    }

    // zip code

    public abstract Object value();

    public abstract boolean isOld();

    public abstract boolean isNew();

    public abstract Object oldZipCodeIfNone(Supplier<Object> aNoneBlock);

    public abstract Object newZipCodeIfNone(Supplier<Object> aNoneBlock);

    // zip code - protected

    protected String code() {
        return code;
    }

    // validation

    protected abstract boolean isValidCode(String aCode);

    protected abstract String invalidCodeErrorDescription();

    // validation - private

    private void assertIsValidCode(String aCode) {
        if (!isValidCode(aCode)) throw new RuntimeException(invalidCodeErrorDescription());
    }
}

File: experiments/020-PythonVsJava/cells/python-pytest_1-Empty_free/2022-1c-recuperatorio-parcial1/20260921-203929-73794/analysis/own-tests-project/src/customerimporter/zip_code.py
from abc import abstractmethod

from customerimporter.imported_value import ImportedValue


class ZipCode(ImportedValue):

    # instance creation

    @classmethod
    def for_code(cls, an_imported_zip_code):
        return cls.subclass_such_that(
            lambda a_zip_code_class: a_zip_code_class.recognizes(an_imported_zip_code)
        ).for_value(an_imported_zip_code)

    # type

    @classmethod
    @abstractmethod
    def recognizes(cls, an_imported_zip_code):
        pass

    def is_new(self):
        return False

    def is_old(self):
        return False

    # zip code

    def new_zip_code_if_none(self, a_none_block):
        return a_none_block()

    def old_zip_code_if_none(self, a_none_block):
        return a_none_block()

    # validation - private

    @classmethod
    def _is_old_zip_code_number(cls, an_imported_text):
        return cls._are_all_digits(an_imported_text) and int(an_imported_text) >= 1000 and int(an_imported_text) <= 9999
